File: 530-MinimumAbsoluteDifferenceInBST/Python/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        # as it is a binary search tree if we
        # browse the tree with an inorder traversal
        # we will see the node in sorted order
        # so we just need to take the min of the previous
        # seeing node and the current node.
        # We can do that with O(1) space like this:
        self.min = float("inf")
        
        def inOrderTraversal(root, prev):
            if not root: return prev
            
            prev = inOrderTraversal(root.left, prev)
            self.min = min(self.min, abs(root.val-prev))
            prev = inOrderTraversal(root.right, root.val)
            return prev
        
        inOrderTraversal(root, -float("inf"))
        return self.min

File: 530-MinimumAbsoluteDifferenceInBST/C++/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    int mini = INT_MAX;
    
public:
    int inOrderTraversal(TreeNode* root, int prev) {
        if (!root) return prev;
        
        int prev_ = inOrderTraversal(root->left, prev);
        
        if (prev_ >= 0) {
            mini = min(mini, abs(root->val - prev_));
        }
        
        // The magic operates here: we passe prev = root->val
        prev_ = inOrderTraversal(root->right, root->val);
        return prev_;
    }
    
    int getMinimumDifference(TreeNode* root) {
        // little trick: we are sure there is no negative value in the tree
        // so pass a negative value as the initial value and don't compute
        // mini if we see a negative value in inOrderTraversal
        inOrderTraversal(root, -1);
        return mini;
    }
};